### PR TITLE
API- Validate capital and reserves balance sheet fields

### PR DESCRIPTION
**Pull request by leahtaylor**
Add api validation for capital and reserves balance sheet fields

Add **@Range** validation to fields
Called up share capital: 0 -99999999
Share premium account: 0 - 99999999
Other reserves: -99999999 - 99999999
Profit and loss account: -99999999 - 99999999

Update Current and Previous period validators to validate ttal shareholder funds total is correct and total shareholder funds equals total net assets

Add **@NotNull** annotation to mandatory total shareholder funds field

Update unit tests to include new fields

SFA-152, SFA-484

### CODE REVIEW

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/PreviousPeriodValidator.java [hunk 1]
```diff
@@ -28,16 +35,37 @@ public Errors validatePreviousPeriod(@Valid PreviousPeriod previousPeriod) {
...
+            Long otherReserves = Optional.ofNullable(capitalAndReserves.getOtherReserves()).orElse(0L);
+            Long profitAndLoss = Optional.ofNullable(capitalAndReserves.getProfitAndLoss()).orElse(0L);
+
+            Long totalShareholderFunds = Optional.ofNullable(capitalAndReserves.getTotalShareholdersFund()).orElse(0L);
+            Long calculatedTotal = calledUpShareCapital + otherReserves + sharePremiumAccount + profitAndLoss;
+            validateAggregateTotal(totalShareholderFunds, calculatedTotal, TOTAL_SHAREHOLDER_FUNDS_PATH, errors);
+
+            Long totalNetAssets = previousPeriod.getBalanceSheet().getOtherLiabilitiesOrAssets().getTotalNetAssets();
+
+            if (totalNetAssets != totalShareholderFunds) {
```
Comment (sroberts2) [inline src/main/java/uk/gov/companieshouse/api/accounts/validation/PreviousPeriodValidator.java]: Object comparison requires .equals

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/CurrentPeriodValidator.java [hunk 2]
```diff
@@ -25,20 +31,42 @@
...
+            Long profitAndLoss = Optional.ofNullable(capitalAndReserves.getProfitAndLoss()).orElse(0L);
 
+            Long totalShareholderFunds = Optional.ofNullable(capitalAndReserves.getTotalShareholdersFund()).orElse(0L);
+            Long calculatedTotal = calledUpShareCapital + otherReserves + sharePremiumAccount + profitAndLoss;
+            validateAggregateTotal(totalShareholderFunds, calculatedTotal, TOTAL_SHAREHOLDER_FUNDS_PATH, errors);
 
-        return errors;
+            // Total shareholder funds must equal total net assets
+            Long totalNetAssets = currentPeriod.getBalanceSheet().getOtherLiabilitiesOrAssets().getTotalNetAssets();
+            if (totalNetAssets != totalShareholderFunds) {
```
Comment (sroberts2) [inline src/main/java/uk/gov/companieshouse/api/accounts/validation/CurrentPeriodValidator.java]: You are comparing objects and need to use `!totalNetAssets.equals(totalShareholderFunds)`

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/PreviousPeriodValidator.java (hunk 1 above)
Comment (leahtaylor) [reply src/main/java/uk/gov/companieshouse/api/accounts/validation/PreviousPeriodValidator.java]: resolved in 39a850d

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/CurrentPeriodValidator.java [hunk 3]
```diff
@@ -25,20 +31,42 @@
...
+
+                addError(errors, shareholderFundsMismatch, TOTAL_SHAREHOLDER_FUNDS_PATH);
+            }
+        }
     }
 
     public void validateTotalCurrentAssets(CurrentPeriod currentPeriod, Errors errors) {
```
[the remark this answers is not in the document]
Comment (leahtaylor) [reply src/main/java/uk/gov/companieshouse/api/accounts/validation/CurrentPeriodValidator.java]: resolved in 5d1fb94

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/PreviousPeriodValidator.java [hunk 4]
```diff
@@ -28,16 +36,37 @@ public Errors validatePreviousPeriod(@Valid PreviousPeriod previousPeriod) {
...
-
             validateTotalCurrentAssets(previousPeriod, errors);
-
+            validateTotalShareholderFunds(previousPeriod, errors);
         }
 
         return errors;
     }
 
+    public void validateTotalShareholderFunds(PreviousPeriod previousPeriod, Errors errors) {
```
[the remark this answers is not in the document]
Comment (leahtaylor) [reply src/main/java/uk/gov/companieshouse/api/accounts/validation/PreviousPeriodValidator.java]: resolved in 5d1fb94

File: src/main/java/uk/gov/companieshouse/api/accounts/validation/CurrentPeriodValidator.java [hunk 5]
```diff
@@ -25,20 +31,42 @@
...
 
+        validateTotalFixedAssets(currentPeriod, errors);
+        validateTotalOtherLiabilitiesOrAssets(currentPeriod, errors);
+        validateTotalCurrentAssets(currentPeriod, errors);
+        validateTotalShareholderFunds(currentPeriod, errors);
+
+        return errors;
+    }
+
+    public void validateTotalShareholderFunds(CurrentPeriod currentPeriod, Errors errors) {
```
[the remark this answers is not in the document]
Comment (leahtaylor) [reply src/main/java/uk/gov/companieshouse/api/accounts/validation/CurrentPeriodValidator.java]: resolved in 5d1fb94

**Comment by leahtaylor** [comment]
@sroberts2 comments addressed :)

**Pull request merged**